packages/angular/ssr/src/routes/router.ts

```diff
@@ -6,86 +6,89 @@
+   */
+  constructor(private readonly routeTree: RouteTree) {}
 
   /**
-   * Indicates whether the route tree needs to be extracted and built.
-   * Initially set to `true` and set to `false` once the routes are built.
+   * Static property to track the ongoing build promise.
    */
-  requiresRoutesToBeBuilt = true;
+  static #extractionPromise: Promise<ServerRouter> | undefined;
 
   /**
-   * Creates an instance of the Router.
+   * Creates or retrieves a `ServerRouter` instance based on the provided manifest and URL.
    *
-   * @param manifest - An instance of `AngularAppManifest` that contains route information.
-   * The manifest provides routing configuration and other details necessary for route management.
+   * If the manifest contains pre-built routes, a new `ServerRouter` is immediately created.
+   * Otherwise, it builds the router by extracting routes from the Angular configuration
+   * asynchronously. This method ensures that concurrent builds are prevented by re-using
+   * the same promise.
+   *
+   * @param manifest - An instance of `AngularAppManifest` that contains the route information.
+   * @param url - The URL for server-side rendering. The URL is needed to configure `ServerPlatformLocation`.
+   * This is necessary to ensure that API requests for relative paths succeed, which is crucial for correct route extraction.
+   * [Reference](https://github.com/angular/angular/blob/d608b857c689d17a7ffa33bbb510301014d24a17/packages/platform-server/src/location.ts#L51)
+   * @returns A promise resolving to a `ServerRouter` instance.
    */
-  constructor(private readonly manifest: AngularAppManifest) {
+  static from(manifest: AngularAppManifest, url: URL): Promise<ServerRouter> {
     if (manifest.routes) {
-      // Initialize the route tree with pre-existing routes from the manifest.
-      this.routeTree = RouteTree.fromObject(manifest.routes);
-      this.requiresRoutesToBeBuilt = false;
-    }
-  }
+      const routeTree = RouteTree.fromObject(manifest.routes);
 
-  private buildPromise: Promise<void> | undefined; // Property to track the ongoing build promise
+      return Promise.resolve(new ServerRouter(routeTree));
+    }
 
-  /**
-   * Builds the route tree.
-   * If a build is already in progress, returns the existing promise.
-   * @param url - The base URL for the routes.
-   * @param html - The HTML document to be used by Angular.
-   * @returns A promise that resolves when the route tree has been built.
-   */
-  buildRoutes(url: URL, html: string): Promise<void> {
-    if (this.buildPromise) {
-      return this.buildPromise;
+    if (ServerRouter.#extractionPromise) {
+      return ServerRouter.#extractionPromise;
     }
 
     // Create and store a new promise for the build process.
     // This prevents concurrent builds by re-using the same promise.
-    this.buildPromise = (async () => {
-      this.routeTree = new RouteTree();
-
+    ServerRouter.#extractionPromise = (async () => {
```

Comment:
Consider: You can drop the `if (ServerRouter.#extractionPromise)` with a nullish coalescing assignment.

```typescript
ServerRouter.#extractionPromise ??= (async () => { /* ... */})();
return ServerRouter.#extractionPromise;
```